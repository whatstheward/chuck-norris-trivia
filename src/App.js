import { useEffect, useState } from "react";
import ChuckNorris from './app/assets/Chuck_Norris.png'
import "./app/sass/styles.css";
import Trivia from './app/components/Trivia';
import Searchbar from "./app/components/Searchbar";

function App() {

  const [trivia, setTrivia] = useState(undefined)
  const [loadingFact, setLoadingFact] = useState('idle')

    
  const fetchChuckFact = () => {
    fetch('https://api.chucknorris.io/jokes/random')
    .then(res => res.json())
    .then(data => {
      setTrivia(data.value)
      setLoadingFact('succeeded')
    })
    .catch(err=> {
      console.log(err)
    })
  }

  useEffect(() => {
      setLoadingFact('pending')
      fetchChuckFact()
    return () => {}
  }, [])

  const handleSubmit = (e, query, resetFn) => {
    e.preventDefault()
    setLoadingFact('pending')
    fetch(`https://api.chucknorris.io/jokes/search?query=${query}`)
    .then(res=>res.json())
    .then(data => {
      if(data.total){
        const randomIdx = data.total > 1 ? Math.floor(Math.random()*data.total)+1 : 0
        setTrivia(data.result[randomIdx].value)
        setLoadingFact('succeeded')
      }else{
        setTrivia('uh-oh. Chuck Norris senses something was wrong.')
        setLoadingFact('rejected')
      }
        resetFn('')
    })
  }
  
  return (
    <div className="App">
      <header>
        <img id='header-img' src={ChuckNorris} alt='Chuck Norris Random Facts'/>
        <h1 className='main-header'>Chuck Norris</h1>
        <h2 className='secondary-header'>Random Facts</h2>
      </header>
      <div className='main-body'>
          <Trivia triviaFact={trivia} loadingFact={loadingFact}/>
        <Searchbar handleSubmit={handleSubmit} loadingFact={loadingFact} fetchChuckFact={fetchChuckFact}/>
      </div>
      <footer>built with <a href='https://api.chucknorris.io' target='blank'>https://api.chucknorris.io</a></footer>
    </div>
  );
}

export default App;
