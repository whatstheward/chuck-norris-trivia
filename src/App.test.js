import * as React from 'react'
import renderer from 'react-test-renderer'
import App from './App'

const resizeWindow = (x,y)=> {
  window.innerWidth = x
  window.innerHeight = y
  window.dispatchEvent(new Event('resize'))
}

describe('App', () => {
  it('renders without crashing', () => {
    const comp = renderer
      .create(<App />)
      .toJSON()
    expect(comp).toMatchSnapshot()
  })
  it('renders image above header in desktop mode', () => {
    const comp = renderer
    .create(<App />)
    .toJSON()
    const node = comp.children[0]
    expect(node.type).toMatch(/header/)
    expect(node.children[0].props.id).toBe('header-img')
  })
  it('does not render image above header in mobile mode', () => {
    resizeWindow(700, 900)
    const comp = renderer
    .create(<App />)
    .toJSON()
    const node = comp.children[0]
    expect(node.children[0].type).toBe('img')
    expect(node.children[0].children).toBe(null)
  })
})