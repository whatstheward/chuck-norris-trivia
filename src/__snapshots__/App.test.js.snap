// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders without crashing 1`] = `
<div
  className="App"
>
  <header>
    <img
      alt="Chuck Norris Random Facts"
      id="header-img"
      src="Chuck_Norris.png"
    />
    <h1
      className="main-header"
    >
      Chuck Norris
    </h1>
    <h2
      className="secondary-header"
    >
      Random Facts
    </h2>
  </header>
  <div
    className="main-body"
  >
    <article
      className="trivia-fact"
    >
      <section>
        <h2>
          Did you know...
        </h2>
        <div>
          <div
            className="dot-pulse"
          />
        </div>
      </section>
    </article>
    <div
      id="search-container"
    >
      <form
        onSubmit={[Function]}
      >
        <h2>
          Search for Facts
        </h2>
        <input
          name="search"
          onChange={[Function]}
          type="text"
          value=""
        />
        <section>
          <button
            disabled={true}
            type="submit"
          >
            Search
          </button>
          <button
            onClick={[Function]}
            type="button"
          >
            Random
          </button>
        </section>
      </form>
    </div>
  </div>
  <footer>
    built with 
    <a
      href="https://api.chucknorris.io"
      target="blank"
    >
      https://api.chucknorris.io
    </a>
  </footer>
</div>
`;

exports[`App should display image above header in desktop mode 1`] = `
<div
  className="App"
>
  <header>
    <img
      alt="Chuck Norris Random Facts"
      id="header-img"
      src="Chuck_Norris.png"
    />
    <h1
      className="main-header"
    >
      Chuck Norris
    </h1>
    <h2
      className="secondary-header"
    >
      Random Facts
    </h2>
  </header>
  <div
    className="main-body"
  >
    <Trivia
      loadingFact="idle"
    />
    <Searchbar
      fetchChuckFact={[Function]}
      handleSubmit={[Function]}
      loadingFact="idle"
    />
  </div>
  <footer>
    built with 
    <a
      href="https://api.chucknorris.io"
      target="blank"
    >
      https://api.chucknorris.io
    </a>
  </footer>
</div>
`;
